### ISSUE TITLE
repr of twisted.protocols.amp.AMP is confusing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3695]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3695|
|Type|defect|
|Created|2009-03-21 16:31:57Z|
|Branch|https://github.com/twisted/twisted/tree/amp-repr-3695|

Someone asked why they have _None_ in the repr of one of their `AMP` instances: `<AuctionProtocol None at 0x35a9a50>`.  _None_ here is the _inner protocol_, a feature of AMP which many people will probably never use.  It would be better to label this value, and perhaps to omit it when there is no inner protocol.

Attachments:

* [amp.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/2d9a359c76c27e35118c98683eda4f2c0de18fd8.patch) (620 bytes) - added by cary on 2009-03-30 21:11:51Z - 
* [3695_unittests.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/981c4fe9d92bba9210d48e6608e42425e6309288.patch) (1005 bytes) - added by SeanMcQ on 2009-04-01 16:49:58Z - Unit tests for patch 3695.
* [3695_unittests2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/52fc0ee8243e5bae1274ae275ea150f942453e7d.patch) (1050 bytes) - added by SeanMcQ on 2009-04-01 17:09:47Z - Unit tests for patch 3695.  Actually testing reported defect.
* [3695_complete.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/7940927dac219836ff6078200abc228cf84f1d27.patch) (1815 bytes) - added by SeanMcQ on 2009-04-01 18:30:02Z - Updated unit tests as per review.  Updated amp.py patch to use if statement for clarity and to use naming standards.
* [3695_complete.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/7efdcacd26cf7abe0fe4239303b7ad8cc74abff3.patch) (1816 bytes) - added by SeanMcQ on 2009-04-01 18:37:31Z - 
* [3695_complete.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/4a0e378716c89e3d365a9d7c3fbfb1d3d9dfe93a.patch) (1808 bytes) - added by SeanMcQ on 2009-04-01 19:09:37Z - 
* [3695_complete.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/6d35cc31bd60283dda519425388b03f7e40337a5.patch) (1808 bytes) - added by SeanMcQ on 2009-04-01 19:10:19Z - 
* [3695_complete.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/d9814e1aacbacc5e627e4872f01a61d5c4887693.patch) (2420 bytes) - added by SeanMcQ on 2009-04-01 20:37:03Z - 
* [3695_complete.6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/829db1eeacd5298ca10fd5566969ba408f4fc850.patch) (2383 bytes) - added by SeanMcQ on 2009-04-01 20:39:00Z - 
* [3695_complete.7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/752/752dce478b2119163d588c861384927c6194ec37/1e30644f6dc1e1157a9d4aac1a20ffa5ce5e0251.patch) (2921 bytes) - added by SeanMcQ on 2009-04-01 21:10:09Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3695 3695
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_amp_repr_3695 branches/amp-repr-3695
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp__repr__easy amp, repr, easy
time__1237653117000000 1237653117000000
changetime__1239674386000000 1239674386000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="cary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|cary removed owner|
|-|-|

If innerProtocol is None:
<Foo at 0xb8e950>

If innerProtocol is set:
<Foo inner<Bar at 0xb8e8f0> at 0xb8e950>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to cary|
|-|-|

New repr looks good to me.  It needs unit tests, though.  Also, `inner_proto` should be `innerProto`.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_4"></a>|@dreid set owner to SeanMcQ|
|-|-|

Ok, your tests should use assert_, there is an assertIn which would be better for this.

Also the repr implementation would probably be clearer if you just used an if statement.

and inner_repr should be innerRepr (see the twisted coding standard for variable naming conventions)

the % operator should always be passed a tuple or a dict.

so it should be ' inner%s' % (self.innerProtocol,)  Also you might want a space between inner and the %s.

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|SeanMcQ commented|
|-|-|

Regression in previous patch.  Output was of the format <AMPinner...> when it was expected to be <AMP inner...>.  3695_complete.2.patch updates to correct this error.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_7"></a>|@dreid commented|
|-|-|

Ok, there is now trailing whitespace in your test methods and the docstring for test_innerProtocolNotInRepr should be wrapped after "innerProtocol".

The string formatting should probably use %r so you definitely get the __repr__ result of the object and not the __str__.

Also there is trailing whitespace on the return line of AMP.__repr__

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|SeanMcQ commented|
|-|-|

Updated per review.

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|SeanMcQ removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_10"></a>|@dreid set owner to SeanMcQ|
|-|-|

Do complete string comparison to check for the repr.

You can do 
```
self.assertEquals(repr(a), "<AMP inner <TestProto None 0x%x> 0x%x>" % (
    id(otherProto),
    id(a)))
```

And that would be better.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_11"></a>|@dreid commented|
|-|-|

Ok, the only comment I have now is that your assertion doesn't need to be on two lines to be readable in the second test.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_12"></a>|@dreid commented|
|-|-|

^- ignore that.

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|SeanMcQ removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun|
|-|-|

  1. the `if self.innerProtocol:` check will "fail" if the protocol has a `__nonzero__` that returns False.  A silly thing to do, but it's easy to catch this kind of thing and deal with it, so we should.  `if self.innerProtocol is not None:` is generally the form of the right solution.
  1. The repr is implemented in terms of `id` (it probably was in trunk already, I didn't check).  This will produce ugly results sometimes, when the unsigned interpretation of the id is greater than 2 ** 31 - it comes out as a negative Python integer.  `twisted.python.util.unsignedID` works around this problem.  (Likewise the test needs to be updated as well).
  1. `TestProto`'s ivars aren't quite formatted right.  _@ivar foo: bar_ is correct - you forgot the _:_.
  1. Even when interpolating a single value, it's best to consistently use a tuple.  `"%s" % bar` is better as `"%s" % (bar,)`.  The ',' is necessary to actually make it a tuple.  `"%s" % (bar)` is the same as the first form, but with unnecessary grouping.  In the case of `id` (or `unsignedID` as I am going to change it to be), the interpolated value is always an integer so this doesn't matter, but using it consistently is a good way to make sure that when it does matter (the case where the interpolated value _might_ be a tuple but you don't know), it gets done right. :)

Otherwise, looks great.  These are minor issues.  I'm going to fix them and put the result into a branch for someone else to take a quick look at.  Thanks! :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [26626]) Branching to 'amp-repr-3695'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [26627]) Apply Rotund's repr fix patch, plus some modifications fixing review feedback

refs [#3695](https://github.com/twisted/twisted/issues/3695)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

Okay, addressed my own review feedback. ;)  Someone else please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @exarkun|
|-|-|

Looks fine to me - merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `closed`|
|-|-|

(In [26697]) Merge amp-repr-3695

Author: SeanMcQ, exarkun
Reviewer: exarkun, glyph
Fixes: [#3695](https://github.com/twisted/twisted/issues/3695)

Change the repr of `twisted.protocols.amp.AMP` instances to exclude the
inner protocol if one does not exist and to be more explicit about what
it is if one does exist.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
